unified_planning/io/pddl_reader.py

```diff
@@ -373,40 +388,47 @@ def _add_effect(
                 for e in exp:
                     to_add.append((e, cond))
             elif op == "when":
-                cond = self._parse_exp(problem, act, types_map, {}, exp[1])
+                cond = self._parse_exp(problem, act, types_map, {}, exp[1], assignments)
                 to_add.append((exp[2], cond))
             elif op == "not":
                 exp = exp[1]
                 eff = (
-                    self._parse_exp(problem, act, types_map, {}, exp),
+                    self._parse_exp(problem, act, types_map, {}, exp, assignments),
                     self._em.FALSE(),
                     cond,
                 )
                 act.add_effect(*eff if timing is None else (timing, *eff))  # type: ignore
             elif op == "assign":
                 eff = (
-                    self._parse_exp(problem, act, types_map, {}, exp[1]),
-                    self._parse_exp(problem, act, types_map, {}, exp[2]),
+                    self._parse_exp(problem, act, types_map, {}, exp[1], assignments),
+                    self._parse_exp(problem, act, types_map, {}, exp[2], assignments),
                     cond,
                 )
                 act.add_effect(*eff if timing is None else (timing, *eff))  # type: ignore
             elif op == "increase":
                 eff = (
-                    self._parse_exp(problem, act, types_map, {}, exp[1]),
-                    self._parse_exp(problem, act, types_map, {}, exp[2]),
+                    self._parse_exp(problem, act, types_map, {}, exp[1], assignments),
+                    self._parse_exp(problem, act, types_map, {}, exp[2], assignments),
                     cond,
                 )
                 act.add_increase_effect(*eff if timing is None else (timing, *eff))  # type: ignore
             elif op == "decrease":
                 eff = (
-                    self._parse_exp(problem, act, types_map, {}, exp[1]),
-                    self._parse_exp(problem, act, types_map, {}, exp[2]),
+                    self._parse_exp(problem, act, types_map, {}, exp[1], assignments),
+                    self._parse_exp(problem, act, types_map, {}, exp[2], assignments),
                     cond,
                 )
                 act.add_decrease_effect(*eff if timing is None else (timing, *eff))  # type: ignore
+            elif op == "forall":
+                assert isinstance(
+                    exp, ParseResults
+                )  # NOTE not sure if we always have a ParseResult here, but this implementation assumes it
```

Comment:
@alvalentini Here, by typing it could also be a string, but I don't know how to handle it

Reply:
Inside that if statement you are sure it is a `ParseResults`